Bluetooth: Controller: Initial broadcast ISO Tx data path

cvinayak opened the pull request:
Initial work on adding ISO Tx data path for ISO Broadcast.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Add Synchronized Receiver board specific conf files

Add Synchronized Receiver board specific conf files.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Add ISO Broadcaster board specific conf files

Add ISO Broadcaster board specific conf files.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix missing ll_adv_cmds_set

Fix to include ll_adv_cmds_set when broadcast role is not
enabled in the Controller but advertising extensions is
enabled (Extended Scanning is used).

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Initial broadcast ISO Tx data path

Initial work on adding ISO Tx data path for ISO Broadcast.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Test ISO data send

Added implementation to test ISO data send in Controller.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Add define for data overflow type ISO data

Add define for data overflow type ISO data.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Pass link_type to data_buffer_overflow

Refactored implementation to pass link type to data buffer
overflow function.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ISO Broadcast and Sync related cond. compile

Fix ISO Broadcast and ISO Synchronized Receiver related
conditional compilation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Broadcast ISO Tx ack on event close

Improvement to generate Tx ack early when the event close
compared to being done at the start of the next event.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot commented:
You beat me to it ;-) Will review Friday morning

cvinayak commented:
> You beat me to it ;-) Will review Friday morning

No, this PR is just for your reference, we take the best of the two. You send yours based on the email you sent, we walk-through and have a formal PR to add the Tx data flow.

mtpr-ot commented on subsys/bluetooth/controller/hci/hci_driver.c:
```diff
@@ -658,6 +659,22 @@ static int acl_handle(struct net_buf *buf)
 }
 #endif /* CONFIG_BT_CONN */
 
+#if defined(CONFIG_BT_CTLR_ADV_ISO)
```
|| defined(CONFIG_BT_CTLR_CONN_ISO)

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/ull.c:
```diff
@@ -842,6 +842,13 @@ uint8_t ll_rx_get(void **node_rx, uint16_t *handle)
+	if (cmplt) {
+		return cmplt;
+	}
+#endif /* CONFIG_BT_CTLR_ADV_ISO || CONFIG_BT_CTLR_CONN_ISO */
```
Don't we need to make sure that only the last PDU counts? I assume that LLID in BIS PDU indicates if this is the last one. Otherwise one SDU could generate more than one complete.

cvinayak replied:
In this PR without ISOAL Tx module, one SDU = one PDU. But I agree, unless ISOAL Tx needs to know each PDU which I guess so for framed PDUs (and may be for unframed)?, only END PDUs could generate the num complete.

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/ull_iso.c:
```diff
@@ -88,14 +95,19 @@ static void iso_rx_demux(void *param);
 #if defined(CONFIG_BT_CTLR_ADV_ISO) || defined(CONFIG_BT_CTLR_CONN_ISO)
-static MFIFO_DEFINE(iso_tx, sizeof(struct lll_tx),
-		    CONFIG_BT_CTLR_ISO_TX_BUFFERS);
-
```
I didn't get rid of the iso_tx mfifo, but kept the whole demuxing path as for ACL. But I can see that we don't actually need it, so great optimization!

cvinayak committed:
Bluetooth: Controller: Add define for Connected ISO Stream handle base

Add define for Connected ISO Stream handle base.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix Extended Advertising conditional compiles

Fix Extended Advertising conditional compiles in header files.
Do not conditional compile declarations, definitions of
functions in source files only be compiled out.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mtpr-ot reviewed (commented):
I think this looks very good. I'm going to create my PR for ISO TX data path very soon, which re-uses the ACL link pool instead of creating separate pool and code. Let's see if it's perhaps better to keep it clean like in this PR.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_iso.c:
```diff
@@ -551,6 +576,55 @@ int ull_iso_reset(void)
+	tx = MFIFO_DEQUEUE_GET(iso_ack);
+	if (tx) {
+#if defined(CONFIG_BT_CTLR_ADV_ISO)
+		*handle = tx->handle + BT_CTLR_ADV_ISO_STREAM_HANDLE_BASE;
```
updated.

cvinayak commented:
@mtpr-ot could I get your approval on this PR, and you can send your changes to override on anything here that need for your requirements?

mtpr-ot commented:
> @mtpr-ot could I get your approval on this PR, and you can send your changes to override on anything here that need for your requirements?

Sure, I'm OK with this. My stuff in #43121 will be a delta, and require a new PR anyway when ISOAL TX is merged.

mtpr-ot reviewed (approved):
Approved as first implementation. ISOAL additions and suggested optimizations coming up.
